Nico/add more content + Latex (#27)
dcSpark/milkomeda-documentation: docs/algorand/how-it-works/data-availability.mdx

With the merkle root of the batch on-chain, the batcher now has 6 blocks to post all of the batch data. The batcher will post one or more grouped Algorand transactions where each group of transactions has a rollup grouping + merkle proof in the transaction notes. (Easiest way to do this will be to have the 1st algorand transaction in the group hold the merkle proof, and the rest of the 15 hold the actual rollup grouping)

During this phase if any batch proposals are submitted, they are simply ignored. If the batcher fails to post all of the batch data within the next 6 blocks, then the batch is considered invalid and ignored (this moves the rollup back to the batch proposal phase).

The rollup nodes deconstruct all of the Algorand transactions over the next 6 blocks to unpack the rollup groupings + their merkle proofs to verify and recreate the merkle tree of the batch. If the batcher posts all of the rollup groupings + merkle proofs, then the rollup nodes can verify the merkle tree is correct, and thus the batch is considered to be available to all of the rollup nodes. 

Now that the batch is verified to be available, it then moves on to batch validation.

### Batch Validation

> The rollup node now takes the batch which has been verified to be available, and validates if the transactions in the batch are all valid (this happens 100% off chain, not on Algorand).

Author: yeah agree

In either case, whether the batch is applied and the rollup progresses forward, or if the batch is ignored leaving the rollup state unchanged, we move back into the batch proposal phase where a new batch has the potential to be added to the rollup history.

Of note, all of this logic is encoded off-chain in the rollup node/auxillary software. This defined protocol is how all the rollup nodes update the state of the rollup itself locally. Anyone can spin up their own rollup node and re-read the Algorand history to deterministically arrive at the same state.

## Advanced Merkelized Batch Proposal Approach

The basic merkelized batch proposal approach allows us to get past the 16kb limit, however only works in a context where there is a single batcher (or fully trusted set of batchers going in order). In this advanced variant, we will be upgrading the batch proposal phase with some more complex dynamics to allow for an open set of batchers to exist who may be adversarial and try to cause problems.

> In this advanced approach, we will be providing all whitelisted batchers (which will likely be a subset of the bridge validators, who eventually will be required to stake assets) the opportunity to be selected as the batch submitter to extend the rollup history.

Reviewer: probably we need to modify this line since the architecture has changed a bit

After the previous batch has either been finalized or ignored (if invalid), we return back to the batch proposal phase. Any of the batchers can now propose a new batch by posting the merkle root of it on-chain. Once a batch proposal is posted by one batcher, all other batchers have 2 blocks to submit their own batch proposals.

After the 2 blocks have passed, one of the batch proposals is pseudo-randomly selected via using the hash of the previous latest batch as a seed to deterministically choose one. The submitter of the batch proposal now is required to post the data of the batch as the protocol moves into the batch availability phase.

The batch availability phase proceeds just as expected, wherein the selected batch submitter must post all of the groupings + merkle proofs of their batch in the next 6 blocks. If they fail to do this (or if during batch validation the batch is invalid), the batch is ignored and the protocol returns to the batch proposal phase.

> Depending on the complexity of the implementation, if specific batchers repeatedly keep failing to submit the data of batches they submit, it would be wise to stake slash them for attempting to DoS the system or simply failing to ensure their software is working properly.

Reviewer: ```suggestion
Depending on the complexity of the implementation, if specific batchers repeatedly keep failing to submit the data of batches they submit, it would be wise to stake slash them for attempting to DoS the system or for simply failing to ensure their software is working properly.
```

## Basic and Advanced Merkelized Batch Proposal Throughput

> In this section we will take a look at some rough estimates of the potential TPS of both the basic and advanced merkelized batch proposal approaches assuming we fill 100% of space of all of the blocks during the batch availability phase. Do note, according to (this post)[https://www.algorand.com/resources/algorand-announcements/algorand-2021-performance] by Silvio Micali, the block size will increase 5x and the block time will decrease from 4.5s -> 2.5s. We will take a look at the potential TPS for both current and future circumstances.

Reviewer: ```suggestion
In this section, we will take a look at some rough estimates of the potential TPS of both the basic and advanced merkelized batch proposal approaches assuming we fill 100% of space of all of the blocks during the batch availability phase. Do note, according to (this post)[https://www.algorand.com/resources/algorand-announcements/algorand-2021-performance] by Silvio Micali, the block size will increase 5x and the block time will decrease from 4.5s -> 2.5s. We will take a look at the potential TPS for both current and future circumstances.
```

**\# Of Txs Possible In Single Batch:** 10,680 txs  
**Algorand Rollup Throughput:** 338 txs/second  

### Advanced Merkelized Batch Proposal Approach
**Batch Proposal Phase Length:** 3 Block  
**Batch Availability Phase Length:** 6 Blocks  
**Total Duration Of Batch Posting:** 9 Blocks  

**\# Of Txs Possible In Single Batch:** 10,680 txs  
**Algorand Rollup Throughput:** 263 txs/second  

### Updated Algorand Throughput
**Average # Of Rollup Txs Per KB:** 1.78 txs/kb  
**Algorand Block Time:** 2.5s  
**Algorand Block Size:** 5mb  

### Basic Merkelized Batch Proposal Approach
**Batch Proposal Phase Length:** 1 Block  
**Batch Availability Phase Length:** 6 Blocks  
**Total Duration Of Batch Posting:** 7 Blocks  

**\# Of Txs Possible In Single Batch:** 53,400 txs  
**Algorand Rollup Throughput:** 3,073 txs/second  

### Advanced Merkelized Batch Proposal Approach
**Batch Proposal Phase Length:** 3 Block  
**Batch Availability Phase Length:** 6 Blocks  
**Total Duration Of Batch Posting:** 9 Blocks  

**\# Of Txs Possible In Single Batch:** 53,400 txs  
**Algorand Rollup Throughput:** 2,373 txs/second  

## Data Availability (v1) Conclusion

> The naive transaction note approach is extremely limited due to the 16kb limit. A max TPS of 6.33 is insufficient for our needs and would likely cause us to take a reputational hit even if it was a first step due to how limited the throughput would be.

Reviewer: ```suggestion
The naive transaction note approach is extremely limited due to the 16kb limit. A max TPS of 6.33 is insufficient for our needs, and thus was never actually implemented.
```

Using the merkelized batch proposal approach, the max TPS is between 263 to 3073 txs/second (depending on which variant, and before or after Algorand improves their base throughput). This represents a 41.5x to 485.5x improvement over the naive approach.

> For Milkomeda we will likely start out with the basic merkelized batch proposal approach, and move towards the advanced approach over time. The basic approach provides us with the majority of the benefits, however is simpler and as such is more centralized as a tradeoff.

Reviewer: ```suggestion
Milkomeda started with the basic merkelized batch proposal approach, and will move towards the advanced approach over time. The basic approach provides it with the majority of the benefits, however is simpler and as such is more centralized as a tradeoff.
```
